### PR TITLE
capture all terminators and quotes in the sentence

### LINKED ISSUE #116: Add an ability to capture all punctuation at the end of a sentence
Currently, if a sentence ends with `？！”`, for example, Yomi will only capture the question mark. Would be nice if it could capture all symbols that are specified as sentence terminators until there aren't any.

### PR DESCRIPTION
**Pull request by Scrub1492**
This PR resolves #116 . It should capture all termination characters or quotes if it appears multiple times at the beginning or end of a sentence. When tested on sentences, it behaves the same as the original but grabs all termination characters.

### CODE REVIEW

**Comment by djahandarie** [comment]
Thanks! Could you add an extra comment or two, and/or make the variables a little more descriptive? (This isn't quite your fault, but I think the code is a bit hard to understand here...)

**Review by toasted-nutbread — COMMENTED** [review]
Would be great to get some test coverage for this case, but also as I was looking into this, I realized that document-util tests are broken. I'll fix those.

File: ext/js/dom/document-util.js
Comment (Scrub1492) [reply ext/js/dom/document-util.js:143]: pos1 > 0 is checked in the while loop though? so inside the loop pos1 should never be negative

Comment (Scrub1492) [reply ext/js/dom/document-util.js:130]: `                    while (typeof terminatorInfo !== 'undefined' && terminatorInfo[0] && pos1 > 0) {`

Comment (toasted-nutbread) [reply ext/js/dom/document-util.js:130]: I'm thinking about the case when `pos1 = 1`, which satisfies `pos1 > 0`. Inside the loop, `pos1` is first decremented, which would give `0`. Then `c = text[pos1 - 1]` is effectively `c = text[0 - 1]`

Comment (Scrub1492) [reply ext/js/dom/document-util.js:130]: Oh, well noted
Ill fix those

Comment (toasted-nutbread) [reply ext/js/dom/document-util.js:130]: It's worth noting that this doesn't actually break anything that I can immediately see, but is still probably best to fix. Out-of-bounds array access in general is kind of lax in Typescript/typed JS sadly.

**Comment by toasted-nutbread** [comment]
> Would be great to get some test coverage for this case, but also as I was looking into this, I realized that document-util tests are broken. I'll fix those.

#363 fixes this; once that's merged, rebase off the latest master and tests can be added to [test\data\html\test-document1.html](https://github.com/themoeway/yomitan/blob/master/test/data/html/test-document1.html).

Here is a new example test which I think will be affected by this update:
```html
    <div
        class="test"
        data-test-type="scan"
        data-element-from-point-selector="span"
        data-caret-range-from-point-selector="span"
        data-start-node-selector="span"
        data-start-offset="4"
        data-end-node-selector="span"
        data-end-offset="4"
        data-result-type="TextSourceRange"
        data-sentence-scan-extent="100"
        data-sentence="ありがとございます。"
    >
        <span>ありがとございます。！？ありがとございます。！？</span>
    </div>
```

**Review by djahandarie — APPROVED** [review]
Thanks for making the fixes! LGTM, but let's also wait for @toasted-nutbread's review.

File: ext/js/dom/document-util.js
Comment (toasted-nutbread) [inline ext/js/dom/document-util.js:121]: I'd recommend not creating an object and just using `cursorStart` and `cursorEnd`. Object notation can have different performance characteristics than inline variables.

Comment (toasted-nutbread) [reply ext/js/dom/document-util.js:121]: To provide a bit more description on this, the reason why I'm suggesting this change is because the `cursor` namespace is more cosmetic vs providing semantic utility. And since there at the very least _can_ be a performance difference with objects, esp. with the variables being used as iteration variables, I recommended to simplify.

For some additional context, I can also list a few situations where I _would_ find such an object to be useful:
* It's passed as a parameter to another function
* It's returned
* It's nullable
* It's already a pre-existing class which closely matches the use case
  (example for something unrelated, imagine a `new Point(x, y)`)

File: ext/js/dom/document-util.js
Comment (toasted-nutbread) [inline ext/js/dom/document-util.js:132]: I wasn't going to comment on this since I thought typescript wouldn't behave well with this change, but you can do a minor optimization of these while loops by removing the undefined check on the first loop, since that's guaranteed by the enclosing `if` statement.

```js
while (terminatorInfo[0] && cursorStart > 0) { // <-- Removed here
    --cursorStart;
    if (cursorStart === 0) { break; }
    c = text[cursorStart - 1];
    terminatorInfo = terminatorMap.get(c);
    if (typeof terminatorInfo === 'undefined') { break; } // <-- Moved here
}
```

**Review by djahandarie — APPROVED** [review]

**Pull request merged**
